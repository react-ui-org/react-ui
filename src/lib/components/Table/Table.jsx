import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import Button from '../Button';
import styles from './Table.scss';

export class Table extends React.Component {
  constructor(props) {
    super(props);

    this.renderHeaderCell = this.renderHeaderCell.bind(this);

    this.sortCellStyle = {
      backgroundColor: 'lightgray',
    };
  }

  renderHeaderCell(column) {
    const {
      id,
      sort,
    } = this.props;
    const sortDirection = sort && column.name === sort.column ? sort.direction : 'asc';
    const isSortingActive = sort && column.name === sort.column;

    return (
      <th
        className={isSortingActive ? styles.isTableHeadCellSortingActive : styles.tableHeadCell}
        key={column.name}
        {...(id && { id: `${id}__headerCell__${column.name}` })}
      >
        {sort && column.isSortable && (
          <div className={styles.sortButton}>
            <Button
              beforeLabel={
                sortDirection === 'asc'
                  ? sort.ascendingIcon
                  : sort.descendingIcon
              }
              label={sortDirection}
              labelVisibility="none"
              onClick={() => sort.onClick(column.name, sortDirection)}
              priority="flat"
              {...(id && { id: `${id}__headerCell__${column.name}__sortButton` })}
            />
          </div>
        )}
        {column.label}
      </th>
    );
  }

  renderBodyCell(column, row) {
    const {
      id,
      sort,
    } = this.props;
    const isSortingActive = sort && column.name === sort.column;

    if (column.format) {
      return (
        <td
          className={isSortingActive ? styles.isTableCellSortingActive : styles.tableCell}
          key={`${row.id}-${column.name}`}
          {...(id && { id: `${id}__bodyCell__${column.name}__${row.id}` })}
        >
          {column.format(row)}
        </td>
      );
    }

    return (
      <td
        className={isSortingActive ? styles.isTableCellSortingActive : styles.tableCell}
        key={`${row.id}-${column.name}`}
        {...(id && { id: `${id}__bodyCell__${column.name}__${row.id}` })}
      >
        {row[column.name]}
      </td>
    );
  }

  render() {
    const {
      columns,
      id,
      rows,
    } = this.props;

    return (
      <table id={id} className={styles.table}>
        <thead>
          <tr className={styles.tableHeadRow}>
            {columns.map(this.renderHeaderCell)}
          </tr>
        </thead>
        <tbody>
          {rows.map((row) => (
            <tr key={row.id} className={styles.tableRow}>
              {columns.map((column) => this.renderBodyCell(column, row))}
            </tr>
          ))}
        </tbody>
      </table>
    );
  }
}

Table.defaultProps = {
  id: undefined,
  sort: null,
};

Table.propTypes = {
  /**
   * Table data columns, optionally sortable. The `format` function can be used to process the
   * column data before displaying them.
   */
  columns: PropTypes.arrayOf(PropTypes.shape({
    format: PropTypes.func,
    isSortable: PropTypes.bool,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
  })).isRequired,
  /**
   * ID of the root HTML element. It also serves as base fo nested elements:
   * * `<ID>__headerCell__<COLUMN_NAME>`
   * * `<ID>__headerCell__<COLUMN_NAME>__sortButton`
   * * `<ID>__bodyCell__<COLUMN_NAME>__<ROW_ID>`
   */
  id: PropTypes.string,
  /**
   * Table data rows, each object key must match a column `name`
   */
  rows: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  /**
   * Sorting configuration required to make columns sortable.
   */
  sort: PropTypes.shape({
    ascendingIcon: PropTypes.node.isRequired,
    column: PropTypes.string.isRequired,
    descendingIcon: PropTypes.node.isRequired,
    direction: PropTypes.oneOf(['asc', 'desc']).isRequired,
    onClick: PropTypes.func.isRequired,
  }),
};

export const TableWithContext = withProviderContext(Table, 'Table');

export default TableWithContext;
