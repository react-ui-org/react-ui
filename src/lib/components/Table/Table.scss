@use "../../styles/tools/transition";
@use "settings";

.table {
    width: 100%;
    border-collapse: collapse;
}

.tableRow,
.tableHeadRow {
    @include transition.add((background-color));
}

.tableRow {
    background-color: settings.$background-color;

    &:hover {
        background-color: settings.$hover-background-color;
    }
}

.tableHeadRow {
    background-color: settings.$head-background-color;

    &:hover {
        background-color: settings.$head-background-color;
    }
}

.tableCell,
.tableHeadCell,
.isTableCellSortingActive,
.isTableHeadCellSortingActive {
    padding: settings.$cell-padding-y settings.$cell-padding-x;
    text-align: left;
    border-bottom: settings.$border-width solid settings.$border-color;
}

.tableHeadCell {
    font-weight: settings.$head-font-weight;
    border-bottom-width: 2px;
}

.isTableCellSortingActive,
.isTableHeadCellSortingActive {
    background-color: settings.$sorted-background-color;
}

.sortButton {
    display: inline-block;
    margin-right: settings.$cell-padding-x;
}
