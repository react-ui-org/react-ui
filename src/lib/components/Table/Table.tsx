import PropTypes from 'prop-types';
import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { TableHeaderCell } from './_components/TableHeaderCell';
import { TableBodyCell } from './_components/TableBodyCell';
import styles from './Table.scss';

export const Table = ({
  columns,
  id,
  rows,
  sort,
  ...restProps
}) => (
  <table
    {...transferProps(restProps)}
    className={styles.table}
    id={id}
  >
    <thead>
      <tr className={styles.tableHeadRow}>
        {columns.map((column) => (
          <TableHeaderCell
            column={column}
            key={column.name}
            sort={sort}
            {...(id && { id: `${id}__headerCell__${column.name}` })}
          />
        ))}
      </tr>
    </thead>
    <tbody>
      {rows.map((row) => (
        <tr key={row.id} className={styles.tableRow}>
          {columns.map((column) => (
            <TableBodyCell
              format={column.format}
              isSortingActive={sort && column.name === sort.column}
              key={`${row.id}-${column.name}`}
              value={row[column.name]}
              {...(id && { id: `${id}__bodyCell__${column.name}__${row.id}` })}
            />
          ))}
        </tr>
      ))}
    </tbody>
  </table>
);

Table.defaultProps = {
  id: undefined,
  sort: null,
};

Table.propTypes = {
  /**
   * Table data columns, optionally sortable. The `format` function can be used to process the
   * column data before displaying them.
   */
  columns: PropTypes.arrayOf(PropTypes.shape({
    format: PropTypes.func,
    isSortable: PropTypes.bool,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
  })).isRequired,
  /**
   * ID of the root HTML element. It also serves as base for nested elements:
   * * `<ID>__headerCell__<COLUMN_NAME>`
   * * `<ID>__headerCell__<COLUMN_NAME>__sortButton`
   * * `<ID>__bodyCell__<COLUMN_NAME>__<ROW_ID>`
   */
  id: PropTypes.string,
  /**
   * Table data rows, each object key must match a column `name`
   */
  rows: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  /**
   * Sorting configuration required to make columns sortable.
   */
  sort: PropTypes.shape({
    ascendingIcon: PropTypes.node.isRequired,
    column: PropTypes.string.isRequired,
    descendingIcon: PropTypes.node.isRequired,
    direction: PropTypes.oneOf(['asc', 'desc']).isRequired,
    onClick: PropTypes.func.isRequired,
  }),
};

export const TableWithGlobalProps = withGlobalProps(Table, 'Table');

export default TableWithGlobalProps;
