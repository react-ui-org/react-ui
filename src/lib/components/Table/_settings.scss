@use "sass:map";
@use "../../styles/theme/borders";
@use "../../styles/theme/colors";
@use "../../styles/theme/typography";
@use "../../styles/tools/spacing";

$cell-padding-x: spacing.of(3);
$cell-padding-y: spacing.of(1);
$border-width: borders.$width;
$border-color: map.get(colors.$grays, gray-100);
$background-color: colors.$white;
$head-background-color: colors.$white;
$head-font-weight: map.get(typography.$font-weight-values, bold);
$hover-background-color: map.get(colors.$grays, gray-50);
$sorted-background-color: map.get(colors.$grays, gray-50);
