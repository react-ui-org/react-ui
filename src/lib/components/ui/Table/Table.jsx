import PropTypes from 'prop-types';
import React from 'react';
import Button from '../Button';
import styles from './Table.scss';

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.renderHeaderCell = this.renderHeaderCell.bind(this);

    this.sortCellStyle = {
      backgroundColor: 'lightgray',
    };
  }

  renderHeaderCell(column) {
    const {
      id,
      sort,
    } = this.props;
    const sortDirection = sort && column.name === sort.column ? sort.direction : 'asc';
    const isSortingActive = sort && column.name === sort.column;

    return (
      <th
        className={isSortingActive ? styles.isTableHeadCellSortingActive : styles.tableHeadCell}
        key={column.name}
        {...(id && { id: `${id}__headerCell__${column.name}` })}
      >
        {sort && column.isSortable && (
          <div className={styles.sortButton}>
            <Button
              clickHandler={() => sort.changeHandler(column.name, sortDirection)}
              beforeLabel={
                sortDirection === 'asc'
                  ? sort.ascendingIcon
                  : sort.descendingIcon
              }
              label={sortDirection}
              labelVisibility="none"
              priority="flat"
              {...(id && { id: `${id}__headerCell__${column.name}__sortButton` })}
            />
          </div>
        )}
        {column.label}
      </th>
    );
  }

  renderBodyCell(column, row) {
    const {
      id,
      sort,
    } = this.props;
    const isSortingActive = sort && column.name === sort.column;

    if (column.format) {
      return (
        <td
          className={isSortingActive ? styles.isTableCellSortingActive : styles.tableCell}
          key={`${row.id}-${column.name}`}
          {...(id && { id: `${id}__bodyCell__${column.name}__${row.id}` })}
        >
          {column.format(row)}
        </td>
      );
    }

    return (
      <td
        className={isSortingActive ? styles.isTableCellSortingActive : styles.tableCell}
        key={`${row.id}-${column.name}`}
        {...(id && { id: `${id}__bodyCell__${column.name}__${row.id}` })}
      >
        {row[column.name]}
      </td>
    );
  }

  render() {
    const {
      columns,
      id,
      rows,
    } = this.props;

    return (
      <div
        className={styles.tableWrapper}
        id={id}
      >
        <table className={styles.table}>
          <thead>
            <tr className={styles.tableHeadRow}>
              {columns.map(this.renderHeaderCell)}
            </tr>
          </thead>
          <tbody>
            {rows.map((row) => (
              <tr key={row.id} className={styles.tableRow}>
                {columns.map((column) => this.renderBodyCell(column, row))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

Table.defaultProps = {
  id: undefined,
  sort: null,
};

Table.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.shape({
    format: PropTypes.func,
    isSortable: PropTypes.bool,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
  })).isRequired,
  id: PropTypes.string,
  rows: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
    ]).isRequired,
  })).isRequired,
  sort: PropTypes.shape({
    ascendingIcon: PropTypes.element.isRequired,
    changeHandler: PropTypes.func.isRequired,
    column: PropTypes.string.isRequired,
    descendingIcon: PropTypes.element.isRequired,
    direction: PropTypes.oneOf(['asc', 'desc']).isRequired,
  }),
};

export default Table;
