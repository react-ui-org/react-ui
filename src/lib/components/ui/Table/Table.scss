@use '../../../styles/tools/transition';
@use 'settings';

// Ignore pascalCase values of `composes` property:
// stylelint-disable value-keyword-case

.table {
  width: 100%;
  border-collapse: collapse;
}

.tableRow {
  @include transition.add((background-color));

  background-color: settings.$background-color;

  &:hover {
    background-color: settings.$hover-background-color;
  }
}

.tableHeadRow {
  composes: tableRow;
  background-color: settings.$head-background-color;

  &:hover {
    background-color: settings.$head-background-color;
  }
}

.tableCell {
  padding: settings.$cell-padding-y settings.$cell-padding-x;
  text-align: left;
  border-bottom: settings.$border-width solid settings.$border-color;
}

.tableHeadCell {
  composes: tableCell;
  font-weight: settings.$head-font-weight;
  border-bottom-width: 2px;
}

.isTableCellSortingActive,
.isTableHeadCellSortingActive {
  background-color: settings.$sorted-background-color;
}

.isTableCellSortingActive {
  composes: tableCell;
}

.isTableHeadCellSortingActive {
  composes: tableHeadCell;
}

.sortButton {
  display: inline-block;
  margin-right: settings.$cell-padding-x;
}
