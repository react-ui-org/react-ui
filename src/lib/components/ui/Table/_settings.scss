@use '../../../styles/settings/borders';
@use '../../../styles/settings/colors';
@use '../../../styles/settings/typography';
@use '../../../styles/tools/spacing';

$cell-padding-x: spacing.of(3);
$cell-padding-y: spacing.of(1);
$border-width: borders.$width;
$border-color: map-get(colors.$grays, gray-100);
$background-color: colors.$white;
$head-background-color: colors.$white;
$head-font-weight: map-get(typography.$font-weight-values, bold);
$hover-background-color: map-get(colors.$grays, gray-50);
$sorted-background-color: map-get(colors.$grays, gray-50);
