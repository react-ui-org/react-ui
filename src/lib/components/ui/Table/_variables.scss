@import '../../../styles/settings/borders';
@import '../../../styles/settings/colors';
@import '../../../styles/settings/typography';
@import '../../../styles/tools/spacing';

$table-cell-padding-x: spacing(3);
$table-cell-padding-y: spacing(1);
$table-border-width: $border-width;
$table-border-color: map-get($colors-grays, gray-100);
$table-background-color: $color-white;
$table-head-background-color: $color-white;
$table-head-font-weight: map-get($typography-font-weight-values, bold);
$table-hover-background-color: map-get($colors-grays, gray-50);
$table-sorted-background-color: map-get($colors-grays, gray-50);
