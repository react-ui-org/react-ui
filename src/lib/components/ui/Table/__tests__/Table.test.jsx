import React from 'react';
import sinon from 'sinon';
import {
  mount,
  shallow,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Table from '../Table';

const columnsData = [
  {
    isSortable: true,
    label: 'Id',
    name: 'id',
  },
  {
    isSortable: true,
    label: 'Name',
    name: 'name',
  },
  {
    format: (row) => row.dateOfBirth.toDateString(),
    label: 'Date of birth',
    name: 'dateOfBirth',
  },
];

const rowsData = [
  {
    dateOfBirth: new Date(2018, 1, 1),
    id: 1,
    name: 'Jan Novak',
  },
];

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<Table
      columns={columnsData}
      rows={rowsData}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<Table
      columns={columnsData}
      id="custom-id"
      rows={rowsData}
      sort={{
        ascendingIcon: <span className="icon" />,
        changeHandler: () => {},
        column: 'id',
        descendingIcon: <span className="icon" />,
        direction: 'asc',
      }}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler() on sorting button click', () => {
    const spy = sinon.spy();
    const component = mount((
      <Table
        columns={columnsData}
        rows={rowsData}
        sort={{
          ascendingIcon: <span className="icon" />,
          changeHandler: spy,
          column: 'id',
          descendingIcon: <span className="icon" />,
          direction: 'asc',
        }}
      />
    ));

    component.find('Button').first().simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.args).toEqual([['id', 'asc']]);
  });
});
