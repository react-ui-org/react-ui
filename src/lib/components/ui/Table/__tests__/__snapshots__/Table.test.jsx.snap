// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="tableWrapper"
>
  <table
    className="table"
  >
    <thead>
      <tr
        className="tableHeadRow"
      >
        <th
          className="tableHeadCell"
          key="id"
        >
          Id
        </th>
        <th
          className="tableHeadCell"
          key="name"
        >
          Name
        </th>
        <th
          className="tableHeadCell"
          key="dateOfBirth"
        >
          Date of birth
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="tableRow"
        key="1"
      >
        <td
          className="tableCell"
          key="1-id"
        >
          1
        </td>
        <td
          className="tableCell"
          key="1-name"
        >
          Jan Novak
        </td>
        <td
          className="tableCell"
          key="1-dateOfBirth"
        >
          Thu Feb 01 2018
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="tableWrapper"
  id="custom-id"
>
  <table
    className="table"
  >
    <thead>
      <tr
        className="tableHeadRow"
      >
        <th
          className="isTableHeadCellSortingActive"
          id="custom-id__headerCell__id"
          key="id"
        >
          <div
            className="sortButton"
          >
            <ForwardRef(withForwardedRef(Button))
              beforeLabel={
                <span
                  className="icon"
                />
              }
              clickHandler={[Function]}
              id="custom-id__headerCell__id__sortButton"
              label="asc"
              labelVisibility="none"
              priority="flat"
            />
          </div>
          Id
        </th>
        <th
          className="tableHeadCell"
          id="custom-id__headerCell__name"
          key="name"
        >
          <div
            className="sortButton"
          >
            <ForwardRef(withForwardedRef(Button))
              beforeLabel={
                <span
                  className="icon"
                />
              }
              clickHandler={[Function]}
              id="custom-id__headerCell__name__sortButton"
              label="asc"
              labelVisibility="none"
              priority="flat"
            />
          </div>
          Name
        </th>
        <th
          className="tableHeadCell"
          id="custom-id__headerCell__dateOfBirth"
          key="dateOfBirth"
        >
          Date of birth
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="tableRow"
        key="1"
      >
        <td
          className="isTableCellSortingActive"
          id="custom-id__bodyCell__id__1"
          key="1-id"
        >
          1
        </td>
        <td
          className="tableCell"
          id="custom-id__bodyCell__name__1"
          key="1-name"
        >
          Jan Novak
        </td>
        <td
          className="tableCell"
          id="custom-id__bodyCell__dateOfBirth__1"
          key="1-dateOfBirth"
        >
          Thu Feb 01 2018
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
